### PR TITLE
feat(java): support Plugin Repository and Enhancing Repository Options

### PR DESCRIPTION
**Pull request by awbdallas**
- Added Plugin Repositories
- Added Maven Repository Policies
- Added Tests

Links:
- https://maven.apache.org/settings.html#repositories

Fixes #

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by awbdallas** [comment]
@mrgrain  Any chance you can take a look?

File: src/java/pom.ts [hunk 1]
```diff
@@ -148,6 +148,32 @@ export interface MavenRepository {
...
+export interface MavenRepositoryPolicy {
+  /*
+   * Is enabled
+   */
+  readonly enabled?: boolean;
+
+  /**
+   * Update Policy
+   */
+  readonly updatePolicy?: "always" | "update" | "never" | `interval:${number}`;
```
Comment (mrgrain) [inline src/java/pom.ts:171]: Needs to be an enum like class, cause type unions don't translate well to java.

File: src/java/pom.ts [hunk 2]
```diff
@@ -148,6 +148,32 @@ export interface MavenRepository {
...
+
+  /**
+   * Checksum Policy
+   */
+  readonly checksumPolicy?: "ignore" | "fail" | "warn";
```
Comment (mrgrain) [inline src/java/pom.ts:176]: Needs to be an enum like class, cause type unions don't translate well to java.

**Review by mrgrain — CHANGES_REQUESTED** [review]
See inline comments. Rest looks fine.

File: src/java/pom.ts [hunk 3]
```diff
@@ -345,6 +402,16 @@ export class Pom extends Component {
...
     };
   }
+
+  private synthPluginRepositories() {
+    if (this.pluginRepositories.length === 0) {
+      return;
+    }
+
+    return {
+      pluginRepositories: { pluginRepositories: this.pluginRepositories },
```
Comment (mrgrain) [inline src/java/pom.ts:412]: I think it should be this?
```suggestion
      pluginRepositories: { pluginRepository: this.pluginRepositories },
```

File: test/java/__snapshots__/pom.test.ts.snap [hunk 4]
```diff
@@ -147,6 +147,35 @@ exports[`addPlugin() with multiple executions 1`] = `
...
+"<?xml version="1.0" encoding="UTF-8"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.acme</groupId>
+    <artifactId>my-artifact</artifactId>
+    <version>1.2.3</version>
+    <packaging>jar</packaging>
+    <name>my-project</name>
+    <pluginRepositories>
+        <pluginRepositories>
```
Comment (mrgrain) [inline test/java/__snapshots__/pom.test.ts.snap:160]: This looks incorrect. See other comment.
```suggestion
    <pluginRepositories>
        <pluginRepository>
```

File: test/java/__snapshots__/pom.test.ts.snap [hunk 5]
```diff
@@ -147,6 +147,35 @@ exports[`addPlugin() with multiple executions 1`] = `
...
+        </pluginRepositories>
+        <pluginRepositories>
+            <id>my-remote-repository</id>
+            <name>Remote Repo</name>
+            <url>https://myserver/repo</url>
+            <layout>default</layout>
+            <policyRepository>
+                <checksumPolicy>fail</checksumPolicy>
+                <updatePolicy>interval:50</updatePolicy>
+            </policyRepository>
```
Comment (mrgrain) [inline test/java/__snapshots__/pom.test.ts.snap:172]: I cannot find a reference for this at all.
Looks like policies are added to `snapshots` and `releases`? 
https://maven.apache.org/pom.html#repositories

Can you help me find the right reference if this is something that does indeed exist?

**Review by mrgrain — CHANGES_REQUESTED** [review]
Apologies this took so long. I've found two issues. See inline comments.

File: test/java/__snapshots__/pom.test.ts.snap (hunk 5 above)
Comment (awbdallas) [reply test/java/__snapshots__/pom.test.ts.snap]: You're absolute right this has been updated!

File: src/java/pom.ts [hunk 6]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
    * The layout of the repository
    */
   readonly layout?: string;
+  /**
+   * Repository Policy
+   */
+  readonly snapshots?: MavenRepositoryPolicy;
+  readonly releases?: MavenRepositoryPolicy;
```
Comment (mrgrain) [inline src/java/pom.ts:156]: Needs a doc string please

File: src/java/pom.ts [hunk 7]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
    * The layout of the repository
    */
   readonly layout?: string;
+  /**
+   * Repository Policy
```
Comment (mrgrain) [inline src/java/pom.ts:153]:
```suggestion
   * Repository Policy for Snapshots
```

File: src/java/pom.ts [hunk 8]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+   */
+  readonly enabled?: boolean;
+
+  /**
+   * Update Policy
+   */
+  readonly updatePolicy?: UpdatePolicy;
+
+  /**
+   * Checksum Policy
```
Comment (mrgrain) [inline src/java/pom.ts:175]:
```suggestion
   * Checksum Policy
   * When Maven deploys files to the repository, it also deploys corresponding checksum files.
```

File: src/java/pom.ts [hunk 9]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+ * @see https://maven.apache.org/settings.html#repositories
+ */
+export interface MavenRepositoryPolicy {
+  /*
+   * Is enabled
+   */
+  readonly enabled?: boolean;
+
+  /**
+   * Update Policy
```
Comment (mrgrain) [inline src/java/pom.ts:170]:
```suggestion
   * Update Policy
   * This element specifies how often updates should attempt to occur. Maven will compare the local POM's timestamp (stored in a repository's maven-metadata file) to the remote.
   * @default UpdatePolicy.DAILY
```

File: src/java/pom.ts [hunk 10]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+  readonly releases?: MavenRepositoryPolicy;
+}
+
+/**
+ * Represents a Maven Repository Policy
+ * @see https://maven.apache.org/settings.html#repositories
+ */
+export interface MavenRepositoryPolicy {
+  /*
+   * Is enabled
```
Comment (mrgrain) [inline src/java/pom.ts:165]:
```suggestion
   * Whether this repository is enabled  for the respective type (`releases` or `snapshots`).
```

File: src/java/pom.ts [hunk 11]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+  readonly updatePolicy?: UpdatePolicy;
+
+  /**
+   * Checksum Policy
+   */
+  readonly checksumPolicy?: ChecksumPolicy;
+}
+
+export class UpdatePolicy {
+  static get always() {
```
Comment (mrgrain) [inline src/java/pom.ts:181]: By the way, these don't need to be getters.
```suggestion
  static get ALWAYS() {
```

File: src/java/pom.ts [hunk 12]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+    return "always";
+  }
+
+  static get update() {
+    return "update";
+  }
```
Comment (mrgrain) [inline src/java/pom.ts:187]:
```suggestion
  static get DAILY() {
    return "daily";
  }
```

File: src/java/pom.ts [hunk 13]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+
+  static get never() {
+    return "never";
+  }
```
Comment (mrgrain) [inline src/java/pom.ts:191]:
```suggestion
  static get NEVER() {
    return "never";
  }
```

File: src/java/pom.ts [hunk 14]
```diff
@@ -148,6 +149,56 @@ export interface MavenRepository {
...
+
+  static interval(num: number) {
```
Comment (mrgrain) [inline src/java/pom.ts:193]:
```suggestion
  /**
   * Updates at an interval of X minutes.
   */
  static interval(minutes: number) {
```

File: src/java/pom.ts [hunk 15]
```diff
@@ -271,6 +324,10 @@ export class Pom extends Component {
     this.repositories.push(repository);
   }
 
+  public addPluginRepository(repository: MavenRepository) {
```
Comment (mrgrain) [inline src/java/pom.ts:327]: Needs a docstring

**Review by mrgrain — CHANGES_REQUESTED** [review]
Code looks mostly good now. Mainly some housekeeping to be done.

**Review by mrgrain — APPROVED** [review]

**Review by mrgrain — APPROVED** [review]

**Pull request merged**